libs/zzz/db/src/Database/DataManagers/CharacterOptManager.ts

```diff
@@ -0,0 +1,196 @@
+      if (!validateTag(tag)) return false
+      if (typeof value !== 'number') return false
+      if (typeof isMax !== 'boolean') return false
+      return true
+    })
+
+    if (optConfigId && !this.database.optConfigs.keys.includes(optConfigId))
+      optConfigId = undefined
+
+    const charOpt: CharOpt = {
+      target,
+      conditionals,
+      bonusStats,
+      statConstraints,
+      optConfigId,
+    }
+    return charOpt
+  }
+
+  // These overrides allow CharacterKey to be used as id.
+  // This assumes we only support one copy of a character opt in a DB.
+  override toStorageKey(key: string): string {
+    return `${this.goKeySingle}_${key}`
+  }
+  override toCacheKey(key: string): CharacterKey {
+    return key.split(`${this.goKeySingle}_`)[1] as CharacterKey
+  }
+  getOrCreate(key: CharacterKey): CharOpt {
+    if (!this.keys.includes(key)) {
+      this.set(key, initialCharOpt(key))
+    }
+    return this.get(key) as CharOpt
+  }
+  setConditional(
+    charKey: CharacterKey,
+    sheet: Sheet,
+    condKey: string,
+    src: Src,
+    dst: Dst,
+    condValue: number
+  ) {
+    this.set(charKey, (charOpt) => {
+      const conditionals = [...charOpt.conditionals]
+      const condIndex = conditionals.findIndex(
+        (c) =>
+          c.condKey === condKey &&
+          c.sheet === sheet &&
+          c.src === src &&
+          c.dst === dst
+      )
+      if (condIndex === -1) {
+        conditionals.push({
+          sheet,
+          src,
+          dst,
+          condKey,
+          condValue,
+        })
+      } else {
+        const cond = conditionals[condIndex]
+        // Check if the value is the same, return false to not propagate the update.
+        if (
+          cond.sheet === sheet &&
+          cond.src === src &&
+          cond.dst === dst &&
+          cond.condKey === condKey &&
+          cond.condValue === condValue
+        )
+          return false
+        cond.sheet = sheet
+        cond.src = src
+        cond.dst = dst
+        cond.condKey = condKey
+        cond.condValue = condValue
+      }
+      return { conditionals }
+    })
+  }
+  setBonusStat(
+    charKey: CharacterKey,
+    tag: Tag,
+    value: number | null, // use null to remove the stat
+    index?: number // to edit an existing stat
+  ) {
+    this.set(charKey, (charOpt) => {
+      const statIndex =
+        index ??
+        charOpt.bonusStats.findIndex((s) => shallowCompareObj(s.tag, tag))
+      const bonusStats = [...charOpt.bonusStats]
+      if (statIndex === -1 && value !== null) {
+        bonusStats.push({ tag, value })
+      } else if (value === null && statIndex > -1) {
+        bonusStats.splice(statIndex, 1)
+      } else if (value !== null && statIndex > -1) {
+        bonusStats[statIndex].value = value
+        bonusStats[statIndex].tag = tag
+      }
+      return { bonusStats }
+    })
+  }
+}
+
+function defOptTarget(key: CharacterKey): CharOpt['target'] {
```

Comment:
i kinda prefer blank default if possible. seems common to miss the target selector and just start optimizing for 'atk' without knowing it

Reply:
probably something in the future, quite a lot of cascading working to allow a "null" state right now

Reply:
https://github.com/frzyc/genshin-optimizer/issues/2802